### PR TITLE
Extract user/address mapping and presence handling from Matrix transport

### LINKED ISSUE #3720: Provide node reachability handling utilities to PFS
## Abstract

To calculate usable routes the PFS needs to know the network reachability state of Raiden nodes.

## Specification

The Raiden MatrixTransport implements reachability state handling via Matrix presence updates.
This functionality needs to be extracted into a reusable utility function that is compatible with the existing Matrix client usage in the PFS.

### PR DESCRIPTION
**Pull request by ulope**
Extract user/address mapping and presence handling from Matrix transport

- Extract the user/address mapping and presence handling into a separate utility class (`UserAddressManager`)
- This allows the services to use this functionality as well (Fixes #3720)
- Complete unit test coverage of the new `UserAddressManager`

Refs: #3124, #3252

### CODE REVIEW

File: raiden/network/transport/matrix/transport.py [hunk 1]
```diff
@@ -997,7 +993,7 @@ def _get_room_for_address(
         if self._stop_event.ready():
             return None
         address_hex = to_normalized_address(address)
-        assert address and address in self._address_to_userids,\
+        assert address and self._address_mgr.is_address_known(address),\
```
Comment (rakanalh) [inline raiden/network/transport/matrix/transport.py]: Could you please separate the assert message into `msg` and get rid of the `\` in this line?

File: raiden/network/transport/matrix/utils.py [hunk 2]
```diff
@@ -28,6 +43,178 @@
...
+USER_PRESENCE_REACHABLE_STATES = {UserPresence.ONLINE, UserPresence.UNAVAILABLE}
+USER_PRESENCE_TO_ADDRESS_REACHABILITY = {
+    UserPresence.ONLINE: AddressReachability.REACHABLE,
+    UserPresence.UNAVAILABLE: AddressReachability.REACHABLE,
+    UserPresence.OFFLINE: AddressReachability.UNREACHABLE,
+    UserPresence.UNKNOWN: AddressReachability.UNKNOWN,
+}
+
+
+class UserAddressManager:
```
Comment (palango) [inline raiden/network/transport/matrix/utils.py]: So in the services we'd instantiate a `UserAddressManager` based on the `GMatrixClient` and provide proper callbacks to update the routing table based on that.

What's the difference between address reachability and user presence though and/or which one should we prefer.


Besides that the PR looks good.

Comment (ulope) [reply raiden/network/transport/matrix/utils.py]: > So in the services we'd instantiate a UserAddressManager based on the GMatrixClient and provide proper callbacks to update the routing table based on that.

Exactly.
One important caveat: In order for presence tracking to occur addresses need to be 'registered' with the `UserAddressManager` by calling the `add_address()` method. 

Question: Would it be useful for you to have an `add_addresses()` method that takes an iterable so you can add all addresses you care about after a restart in one go?

> What's the difference between address reachability and user presence though and/or which one should we prefer.

For our purposes an `address` here is basically equivalent to a Raiden node.
One node with one `address` can be represented by multiple `user`s on multiple homeservers.
In the Raiden transport we need to know about individual user presence in order to react to changes in that (e.g. invite 'new' users for already known addresses into existing communication rooms).

In the services you shouldn't need to care about users, only addresses. That means you also don't need to provide a user presence callback.

I'll add a docstring to the class explaining this.

Comment (palango) [reply raiden/network/transport/matrix/utils.py]: > Question: Would it be useful for you to have an add_addresses() method that takes an iterable so you can add all addresses you care about after a restart in one go?

The participants are stored on a per channel basis, so this wouldn't make a big improvement.

> I'll add a docstring to the class explaining this.

Thanks for the explanation.

**Pull request merged**